src/lib/OS/MemRegion.cc

```diff
@@ -83,39 +76,47 @@ uint64_t MemRegion::updateBrkRegion(uint64_t newBrk) {
   return brk_;
 }
 
-uint64_t MemRegion::addVma(VMA* vma, uint64_t startAddr) {
-  VMA* curr = vm_head_;
-  size_t size = vma->vmSize_;
-  // When starAddr is not 0, search for an available address range
-  // that can hold the new VMA with a starting address that is greater than or
-  // equal to the specified startAddr. The following algorithm retrieves the
-  // last existing VMA object before the address range so that the new VMA can
-  // be linked between two existing VMAs. If no available address range is
-  // found, then the new VMA is allocated at the end of the VMA list.
-  if (startAddr && vm_size_ > 0) {
-    while (curr->vmNext_ != nullptr) {
-      if (curr->vmEnd_ <= startAddr &&
-          curr->vmNext_->vmStart_ >= (startAddr + size)) {
+uint64_t MemRegion::addVma(VMA vma, uint64_t startAddr) {
+  size_t size = vma.vmSize_;
+  auto last = std::prev(VMAlist_->end(), 1);
+  bool allocated = false;
+
+  auto itr = VMAlist_->begin();
+  if (VMAlist_->size() > 0) {
+    // Check if the new VMA can be allocated between mmapStart and the first VMA
+    // in the VMA list.
+    uint64_t effectiveMmapStart = startAddr ? startAddr : mmapStart_;
```

Comment:
The addVma function is called in mmapRegion which does the check you mentioned.

Reply:
Would it not be better suited to be inside the function to avoid possible errors in the future of calling the function directly without doing the manual check beforehand?

Reply:
`mmapRegion` function handles all the prelim checks i.e flags, alignment, addr availability, etc. This function is only responsible for adding the vma to a vma list, I think adding the `startAddr` check would just complicate this function (because then we would check for an error value in mmapRegion, currently addVma will only be called once all other checks have been succeeded). This function is private and cannot be called by an instantiation of memRegion. 
@FinnWilkinson any thoughts on this?

Reply:
I think moving this check into `addVma` would be safer if this is used elsewhere in future
```    
if (!((startAddr >= mmapStart_) && (startAddr + size < mmapEnd_))) {
      std::cout << "[SimEng:MemRegion] Provided address range doesn't exist in "
                   "the mmap range: "
                << startAddr << " - " << startAddr + size << std::endl;
      return -1;
}
```

Reply:
done